qutebrowser/utils/usertypes.py

```diff
@@ -246,6 +247,52 @@ def reset(self):
                      'err_config', 'err_key_config'], is_int=True, start=0)
 
 
+# Where a download should be saved
+class DownloadTarget:
```

Comment:
I agree it's kind of weird in Python (I probably got too much Rust lately :wink:), alternatives include (but not limited to):

* :white_check_mark: Add another argument, like `open_file=False`, instead of the special `OPEN_DOWNLOAD` object (and check with `if sum([fileobj is not None, filename is not None, open_download]) > 1: raise ValueError("At most one of fileobj, filename or open_download may be given")`)
* :white_check_mark:  Use a "normal" enum and extra parameters for the filename/fileobj (e.g. ` download_manager.fetch(target=Target.Fileobj, fileobj=...)`, but that would be the same hassle to check (`if target != Target.fileobj and fileobj is not None: raise ValueError("fileobj= is only supported for target=Fileobj")`
* :negative_squared_cross_mark: Similar to the previous point, separate the enum and the associated data, but put it in an object instead of parameters (like `target = DownloadTarget(); target.type = DownloadTarget.FileObj; target.fileobj = io.BytesIO()`) (IMHO looks more stupid than the current implementation and it needs the same amount of documentation as the "separate arguments"-approach because it just hides those arguments in a single object)
* :negative_squared_cross_mark: Don't expose the ability to open downloads to anything except the prompter, though that would only hide the problem and turn it into an undocumented feature. (IMHO bad because it forbids things as hinting to directly open a download or a `:download --open` command)
* :interrobang: Go full OO and completely decouple the target-saving code from the manager, take it out of DownloadItem, specify an abstract interface resembling a file-like object, implement the code that was taken out of DownloadItem in various concrete classes (`class BinaryDownloadDataToFileObjectSink(AbstractBinaryDownloadDataSink): ...`) and then pass objects of one of those classes to the download manager. :wink: (tongue-in-cheek aside, it could be a serious possibility to define an abstract interface and let different classes implement different download targets). (unsure because it's more effort to write and more verbose, only useful if we really need the extensibility for e.g. plugins)

I just think it's weird to have one argument meaning two things (filename being either a filename or OPEN_DOWNLOAD), and then another argument for another download target. So the idea is to either unify them all (the "enum approach") or separate them all (the... "separation approach"). I think the enum approach is more readable when there are many choices, since it eliminates the need to document that only one of those options may be given, and it doesn't clutter the parameters (especially if later more methods are added), at the cost of a slighly more verbose call-site (and the frankenstein enum).